### ISSUE TITLE
How does hilbert know from the config which ServiceRuntime to use?

### ISSUE THREAD

**Issue by porst17** [hilbert/hilbert-cli#32]
Services and application have an `options` field which is supposed to specify the way the service or application is launched. But how does hilbert know if the given options belong to the `docker-compose`, `flatpak`, `appimage`, `systemd`, ... way of launching things?

Do we need a `type` attribute to specify that?

**Comment by malex984** [comment]
Hm, originally i was counting on using YAML tags `!!ClassName {...}`. But if tags are to be avoided then we can do one of the following instead:
- explicitly introduce `type` field directly into the base concept 
- auto-detect type after looking at fields (unsafe)
- type may depend on the global `Version` (at this stage we only support  Docker-Compose..)
- use type-dependent names instead of `options`, e.g. `docker_runtime_options`, `docker_compose_runtime_options` etc. (ugly)

Any other alternatives?

I am going to introduce a `type` field into both **`ServiceRunTime`** and **`PowerOnMethod`**.

**Comment by malex984** [comment]
Hm, originally i was counting on using YAML tags `!!ClassName {...}`. But if tags are to be avoided then we can do one of the following instead:
 - explicitly introduce `type` field directly into the base concept 
 - auto-detect type after looking at fields (unsafe)
 - type may depend on the global `Version` (at this stage we only support  Docker-Compose..)
 - use type-dependent names instead of `options`, e.g. `docker_runtime_options`, `docker_compose_runtime_options` etc. (ugly)

Any other alternatives?

I am going to introduce a `type` field into both **`ServiceRunTime`** and **`PowerOnMethod`**.

**Comment by malex984** [comment]
Hm, originally i was counting on using YAML tags `!!ClassName {...}`. But if tags are to be avoided then we can do one of the following instead:
 - explicitly introduce `type` field directly into the base concept 
 - auto-detect type after looking at fields (unsafe)
 - type may depend on the global `Version` (at this stage we only support  Docker-Compose..)
 - use type-dependent names instead of `options`, e.g. `docker_runtime_options`, `docker_compose_runtime_options` etc. (ugly)

Any other alternatives?

I am going to introduce a `type` field into both **`ServiceRunTime`** and **`PowerOnMethod`**.

**Comment by elondaits** [comment]
Yes, a `type` attribute. It's the simplest and all the other ideas have problems I can think of.

But it would seem the Station should have a fixed type as well... not just the app... because a station might not have Docker support installed, or a Windows station wouldn't support systemd. But this could wait... I wouldn't worry too much about how to handle different "types" until we have at least two types implemented.

**Comment by porst17** [comment]
I'd say that a profile needs `runtime_types` as well then, e.g. you could set `runtime_types: [ flatpak, docker-compose, systemd ]` to indicate that stations with this profile can run flatpak, docker-compose and systemd apps and services. 

I propose to also add some kind of `native` type (or `exec`) that just runs the specified command with options on the remote host, where the respective program must be installed already. This should serve us well for the special stations with pre-installed software.

For our current client, we need to support at least the types `docker-compose` and `native` (resp. `exec`).

**Comment by malex984** [comment]
I added `type` with https://github.com/hilbert/hilbert-cli/commit/fece7b8130d5f2c48e53bdac70fa86b7779b824b

ps: there were also some renames. Google Document is up-to-date.

**Comment by malex984** [comment]
@porst17 Actually any other services / applications types can be viewed as specializations of a `Native` runtime... Maybe `ServiceRuntime` should actually be that **native** execution variant? 

On the other hand: any `runtime` is more than just a single command execution. 
I would count the following to a general `ServiceRuntime` functionality:
- fetch service/application bundle to the host 
- cleanup previously fetched service/application bundles from the host 
- start service/application in background, while passing some settings into it
- stop/kill service/application which is running on a station
- monitor currently running service/application

Therefore i would postpone implementing such generality until later.

**Comment by malex984** [comment]
@porst17 adding `runtime_types` sounds like an additional constraint for configuration validation. IMO we may add that later on.

**Comment by porst17** [comment]
Regarding `runtime_types`: The other option is to leave it out and assume that the user who write the config knows what he is doing, e.g. if he adds a service of a specific type to a profile he must make sure that the prerequisites are met, e.g. `docker-compose`, `systemd`, ... is installed and properly configured.

**Comment by elondaits** [comment]
It's possible, but the "station groups" (supported_stations) provide a very granular and extensive support for users to indicate which stations are compatible with which applications... so it'd be weird to not specify the one very basic aspect of compatibility which directly concerns and affects us.

I'm OK, though, with leaving this for later. In fact if the station could override the list of "services" from the profile by specifying them on its own config I would leave the whole profile thing for later when we can see how to (perhaps) mix the "extends" and "profile" thing into one single way of grouping similar settings.

**Comment by malex984** [comment]
>  if the station could override the list of "services" from the profile by specifying them on its own config

Although technically easy (e.g. via `client_settings.hilbert_station_services`) IMO it may be very confusing if we still keep the original idea of filtering by profiles in UI.

Moreover since profiles represent disjoint sets of stations => selecting several profile may be used for a **union** of corresponding sets of stations. Although groups are very similar to profiles - they are not disjoint => users may need to filter by selecting several groups to get a group **intersection**. I would suggest to consider adding such a usage into the UI later on.

Note moreover that Stations and Profiles will be presented to users via UI. That is why i added `name` / `description` and `icon` fields to them. 

@porst17 @elondaits **Q**: if (explicitly defined) Groups will be visible to users just like profiles: should they also have fields like `name` / `description` / `icon` in their definitions?

**Comment by malex984** [comment]
Again back to the issue topic: Runtime was not a good concept... in fact **`Service`** itself is an abstract class without a concrete specification of its `type` and corresponding settings may be simply embedded into the Service directly:  see corresponding commit: https://github.com/hilbert/hilbert-cli/commit/ae169e2def56eed88d4b082d76481b3c1d2be160

ps: there were also some minor changes. Google Document is up-to-date.

**Comment by porst17** [comment]
I think @elondaits is right: Profiles and the `extends` mechanism both provide a way to reuse and group information. To me, it looks like profiles can be dropped altogether. The only thing they add to a station is the list of services to run and this could be done by the `extends` mechanism. But the idea of dropping profiles is not well thought through and we should not do it right now but rather in a later iteration of changes to the config or cli (which I am sure will happen).

I suggest to add the field `supported_types` to `Profile` now (`supported_types` would be more in line with `supported_stations` than `runtime_types`). Otherwise, the addition of `type` to `Service` would not make sense either because we are not checking it.

**Comment by malex984** [comment]
Profiles define the order in which to start/stop Services before starting/stopping an Application.
We can treat Service just like Applications: e.g. by adding something like `run_on` of type `Group` (by analogy with `compatibleStations`) to each `Service`. How can we manage dependencies then?

~~ps: I will update Google Doc + ConfigurationDD a bit later.~~

**Comment by malex984** [comment]
New ConfigurationDD: is due to https://github.com/hilbert/hilbert-cli/commit/2b5c8ed8b90713cf93af8a6a1e543d6acc46945d

Google Doc was also updated

**Comment by malex984** [comment]
Design is in the Management Back-end Document. Initial parsing is due to (#34) 
But handling of `Profile::supported_types` will be implemented and tested within (#14).
